api/wasm/rust/src/context.rs

```diff
@@ -0,0 +1,142 @@
+use crate::host::*;
+use lazy_static::lazy_static;
+use std::collections::HashMap;
+use std::ffi::CString;
+use std::os::raw::c_char;
+use std::ptr::null_mut;
+use std::sync::{Arc, Mutex};
+
+pub trait RootContext {
+  fn on_start(&self) -> u32;
+}
+pub trait Context {
+  fn on_create(&self);
+}
+
+pub trait RootContextFactory {
+  fn create(&self) -> Arc<Mutex<dyn RootContext + Sync + Send>>;
+}
+pub trait ContextFactory {
+  fn create(
+    &self,
+    _root_context: Arc<Mutex<dyn RootContext + Sync + Send>>,
```

Comment:
i dont think you need Arc \ mutex here.
regular Rc without mutex should work, as IIRC VMs are instantiated per thread.
even if that wasn't the case i would still go for Arc but without mutex (similar to how envoy filters work today).

Reply:
It may be not required Mutex, but required Arc. Because we use only Rc, we can't guarantee the thread safety of static region variables and rustc can't pass them. It can pass with `thread_local!` macro. But we can't apply `thread_local!` and `lazy_static!` simultaneously.

So we should rewrite it to 
```
Arc<dyn RootContext + Sync + Send>
```